crypto/zk/circuit/circuit.go

```diff
@@ -0,0 +1,278 @@
+	if err != nil {
+		return false, err
+	}
+
+	vKeyVerified, err := checkHash(circuit.VerificationKey, circuit.Config.VerificationKeyHash)
+	if err != nil {
+		return false, err
+	}
+
+	wasmVerified, err := checkHash(circuit.Wasm, circuit.Config.WasmHash)
+	if err != nil {
+		return false, err
+	}
+
+	return zKeyVerified && vKeyVerified && wasmVerified, nil
+}
+
+// checkHash compute the hash of the content provided and compares it with the
+// hash provided as expected result. It returns a boolean with the result of the
+// comparation and with an error.
+func checkHash(content, expected []byte) (bool, error) {
+	if content == nil {
+		return false, fmt.Errorf("no content provided to check")
+	} else if expected == nil {
+		return false, fmt.Errorf("no hash provided to compare")
+	}
+
+	hash := sha256.New()
+	if n, err := hash.Write(content); err != nil {
+		return false, fmt.Errorf("error computing hash function of %s: %w", content, err)
+	} else if n != len(content) {
+		return false, fmt.Errorf("the number of writted bytes does not match with the content provied")
+	}
+
+	return bytes.Equal(hash.Sum(nil), expected), nil
+}
+
+// downloadFile functions perform a GET request to the URL provided and returns
+// the content of the received response. If something fails returns an error.
+func downloadFile(ctx context.Context, fileUrl string) ([]byte, error) {
+	if _, err := url.Parse(fileUrl); err != nil {
+		return nil, fmt.Errorf("error parsing the file URL provided: %w", err)
+	}
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, fileUrl, nil)
+	if err != nil {
+		return nil, fmt.Errorf("error creating the file request: %w", err)
+	}
+
+	res, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return nil, err
+	} else if res.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("error on download file %s: http status: %d", fileUrl, res.StatusCode)
+	}
+
+	defer res.Body.Close()
+	content, err := io.ReadAll(res.Body)
+	if err != nil {
+		return nil, fmt.Errorf("error reading the file content from the http response: %w", err)
+	}
+
+	return content, nil
+}
+
+// storeFile helper function allows to write the file content provided into a
+// new file created at the path provided.
+func storeFile(content []byte, dstPath string) error {
+	if content == nil {
+		return fmt.Errorf("no content provided")
+	} else if _, err := os.Stat(filepath.Dir(dstPath)); err != nil {
+		return fmt.Errorf("destination path parent folder does not exist")
+	}
+
+	fd, err := os.Create(dstPath)
+	if err != nil {
+		return fmt.Errorf("something was wrong creating the artifact file: %w", err)
+	}
+
+	if nBytes, err := fd.Write(content); err != nil {
+		return fmt.Errorf("something was wrong writting the artifact file: %w", err)
+	} else if len(content) != nBytes {
+		return fmt.Errorf("something was wrong writting the artifact file: the length of the provided content does not match with the bytes writted")
```

Comment:
writing

written

also, according to fd.Write() docs, the whole check `len(content) != nBytes` is unnecessary:
`// Write returns a non-nil error when n != len(b).`

Reply:
fixed! I dont know that... :D